### ISSUE TITLE
Implementing Memory-Efficient Adam Optimizer

### ISSUE THREAD

**Issue by rdspring1** [pytorch/pytorch#30478]
## 🚀 Feature
Memory-Efficient Adam Optimizer 
- Implement compressed, adaptive optimizer for training large-scale transformer models

## Motivation

You can get out-of-memory issues when training large-scale transformer models (e.g. BERT, GPT-2, RoBERTa). Since the models are so large, adaptive optimizers (e.g. Adam) can take up too much space. A memory-efficient adaptive optimizer would complement activation checkpointing and mixed-precision training. 

For example, you cannot train a GPT-2 774M parameter model using Nvidia V100 16 GB cards with the standard Adam optimizer, activation checkpointing, and FP16 mixed precision training. 

Tensorflow already has memory-efficient optimizers. See [SM3](https://github.com/google-research/google-research/tree/master/sm3).

## Pitch
[Count-Sketch Optimizers](https://github.com/rdspring1/Count-Sketch-Optimizers) is a compressed adaptive optimizer for PyTorch. It includes results on successfully training BERT-Large models.

Usage Examples:
**Original Adam Optimizer**
```
import torch.optim.Adam as Adam
optimizer = Adam(model.parameters(), betas=(0.9, 0.999))
```
**Compressed RMSprop Optimizer**
```
import torch.optim.compressed_rmsprop as cRMSprop
optimizer = cRMSprop(model.parameters())
```

cc @vincentqb

**Comment by pietern** [comment]
Thanks for posting. This seems good to have indeed.

**Comment by Enealor** [comment]
I implemented a version of [SM3 for PyTorch](https://github.com/Enealor/PyTorch-SM3) after seeing this post. I'm not sure if there is interest in incorporating it into PyTorch, but I thought I would mention it here for @rdspring1. I realized that the Count-Sketch method you describe could be incorporated into SM3 in order to compress the momentum term. In both the TF version and my own, the momentum term is the same size as the gradient, which ruins the memory savings. The method you describe could be used to improve this.
